### PR TITLE
#24 Feat/feature cart

### PR DESCRIPTION
**Pull request by minhng271**
### Feature cart
- Show cart bar
- Close cart bar
- Show cart icon with a badge number 
   - Update total item cart when add / delete item
### Result

https://user-images.githubusercontent.com/49194444/212594120-6adb7c9b-088d-4dc7-bc6c-33a30d3f5ffe.mp4

### CODE REVIEW

File: react-basic-practice/src/components/CartBar/index.jsx [hunk 1]
```diff
@@ -2,57 +2,83 @@ import React from 'react';
...
-];
-
-const CartBar = () => {
-  // const {
-  //   onClose, data,
-  // } = props;
-  const onDeleteCartItem = () => {};
-  const onHandleCloseNav = () => {
-    // onClose(false);
+const CartBar = (props) => {
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/CartBar/index.jsx]: CartSideBar

File: react-basic-practice/src/components/CartBar/index.jsx [hunk 2]
```diff
@@ -2,57 +2,83 @@ import React from 'react';
...
+  const { handleCloseCart } = props;
+  const { listItem, setListItem, updateItemCart } = useCart();
+  const { toast, setToast } = useToast();
+
+  const onDeleteCartItem = (e) => {
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/CartBar/index.jsx]: Event handler should have prefix  `handle`: `handleDeleteCartItem `

File: react-basic-practice/src/components/CartBar/index.jsx [hunk 3]
```diff
@@ -2,57 +2,83 @@ import React from 'react';
...
+    try {
+      const idSelected = e.target.closest('.cart-item').id;
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/CartBar/index.jsx]: Please use this: https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes

File: data/db.json [hunk 4]
```diff
@@ -78,12 +78,6 @@
           "quantity": 3,
           "name": "hal earrings",
           "price": 25
-        },
-        {
-          "idProduct": 1,
```
Comment (MinhNguyen41092) [inline data/db.json:83]: Should use uuid for `ID`.

File: react-basic-practice/src/components/CartBar/index.jsx [hunk 5]
```diff
@@ -2,57 +2,83 @@ import React from 'react';
...
         <div className="list-cart-item">
           {
-          data.map((item) => (
-            <CartItem
-              item={item}
-              handleDeleteCartItem={onDeleteCartItem}
-            />
-          ))
+            listItem.listProducts.map((item) => (
+              <CartItem
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/CartBar/index.jsx]: What prop shold you have when looping?

File: react-basic-practice/src/contexts/CartProvider.jsx [hunk 6]
```diff
@@ -1,21 +1,37 @@
...
+        setListItem(dataCart);
+      } catch {
+        setListItem([]);
+      }
+    };
+
+    getData();
+  }, []);
+
+  const updateItemCart = useCallback(async (data) => {
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/contexts/CartProvider.jsx]: Not the right way to use `useCallback`

Comment (MinhNguyen41092) [reply react-basic-practice/src/contexts/CartProvider.jsx]: Also this method should not be here.

File: react-basic-practice/src/components/CartBar/index.jsx (hunk 5 above)
Comment (minhng271) [reply react-basic-practice/src/components/CartBar/index.jsx]: That's the key, I'll use the id for the key

File: react-basic-practice/src/layouts/Header/index.jsx [hunk 7]
```diff
@@ -1,26 +1,48 @@
...
-      <Button
-        type="button"
-        className="btn-cart"
-        icon={cartButton}
-      />
-    </div>
-  </header>
-);
+      {
+        isOpen && <CartSideBar handleCloseCart={onCloseCart} />
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/layouts/Header/index.jsx]: The rule is when you define prop name it's `on<EventName>`. When you define event handler method it's `handle<EventName>`.

File: react-basic-practice/src/components/CartSideBar/index.jsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                handleDeleteCartItem={handleDeleteCartItem}
+              />
+            ))
+        }
+        </div>
+      </div>
+      {
+        toast.openPopup
+        && (
+        <Toast
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/components/CartSideBar/index.jsx:77]: In the next PR please find a way to call Toast in one place for the whole app.

File: react-basic-practice/src/layouts/Header/index.jsx [hunk 9]
```diff
@@ -1,26 +1,48 @@
...
-      <Button
-        type="button"
-        className="btn-cart"
-        icon={cartButton}
-      />
-    </div>
-  </header>
-);
+      {
+        isOpen && <CartSideBar handleCloseCart={handleCloseCart} />
```
Comment (MinhNguyen41092) [inline react-basic-practice/src/layouts/Header/index.jsx]: onCloseCart={handleCloseCart}

**Review by MinhNguyen41092 — APPROVED** [review]

**Pull request merged**